plugins/optimization-detective/site-health.php

```diff
@@ -0,0 +1,304 @@
+		$body    = wp_remote_retrieve_body( $response );
+		$data    = json_decode( $body, true );
+
+		$required_args = array( 'slug', 'current_etag', 'hmac', 'url', 'viewport', 'elements' );
+		$is_expected   = (
+			400 === $code &&
+			isset( $data['code'], $data['data']['params'] ) &&
+			'rest_missing_callback_param' === $data['code'] &&
+			is_array( $data['data']['params'] ) &&
+			count( array_intersect( $data['data']['params'], $required_args ) ) === count( $required_args )
+		);
+		if ( ! $is_expected ) {
+			$result['status']      = 'critical';
+			$result['label']       = __( 'The Optimization Detective REST API endpoint is unavailable to logged-out users', 'optimization-detective' );
+			$result['description'] = $common_description_html . $error_description_html . '<p>' . wp_kses(
+				sprintf(
+					/* translators: %d is the HTTP status code, %s is the status header description */
+					__( 'The REST API returned with an HTTP status of <code>%1$d %2$s</code>.', 'optimization-detective' ),
+					$code,
+					esc_html( $message )
+				),
+				array( 'code' => array() )
+			) . '</p>';
+
+			if ( isset( $data['message'] ) && is_string( $data['message'] ) ) {
+				$result['description'] .= '<blockquote>' . esc_html( $data['message'] ) . '</blockquote>';
+			}
+
+			$result['description'] .= '<details><summary>' . esc_html__( 'Raw response:', 'optimization-detective' ) . '</summary><pre style="white-space: pre-wrap">' . esc_html( $body ) . '</pre></details>';
+		}
+	}
+	return $result;
+}
+
+/**
+ * Gets the response to an Optimization Detective REST API store request to confirm it is available to unauthenticated requests.
+ *
+ * @since n.e.x.t
+ * @access private
+ *
+ * @param bool $use_cached Whether to use a previous response cached in a transient.
+ * @return array{ response: array{ code: int, message: string }, body: string }|WP_Error Response.
+ */
+function od_get_rest_api_health_check_response( bool $use_cached ) {
+	$transient_key = 'od_rest_api_health_check_response';
+	$response      = $use_cached ? get_transient( $transient_key ) : null;
+	if (
+		(
+			is_array( $response )
+			&&
+			isset( $response['response']['code'], $response['response']['message'], $response['body'] )
+			&&
+			is_int( $response['response']['code'] )
+			&&
+			is_string( $response['response']['message'] )
+			&&
+			is_string( $response['body'] )
+		)
+		||
+		is_wp_error( $response )
+	) {
+		return $response;
+	}
+	$rest_url = get_rest_url( null, OD_REST_API_NAMESPACE . OD_URL_METRICS_ROUTE );
+	$response = wp_remote_post(
+		$rest_url,
+		array(
+			'headers'   => array( 'Content-Type' => 'application/json' ),
+			'sslverify' => false,
+		)
+	);
+
+	// This transient will be used when showing the admin notice with the plugin on the plugins screen.
+	// The 1-day expiration allows for fresher content than the weekly check initiated by Site Health.
+	set_transient( $transient_key, $response, DAY_IN_SECONDS );
```

Comment:
I'm a bit concerned about trying to store a potential `WP_Error` object in a transient. Can we parse that into a raw data array instead to make this safer?

Reply:
Why is it concerning to store a `WP_Error` in a transient? The object gets serialized.

Reply:
Hmm, is there prior art for that in WordPress? I suppose it could work, but have you tested this, both using the DB for transients as well as a persistent object cache?

Reply:
Looking at the `WP_Error` class, for instance I would expect the `$additional_data` property to be missing, since it's not `public`. Granted, that data may not be important for what we're doing here, but it shows the fragility of relying on serialization.

Reply:
As noted above, the `test_od_get_rest_api_health_check_response` test confirms that transients return unserialized `WP_Error` instances.

Reply:
What about persistent object cache? It doesn't look like our unit tests are running with one, so we would need to separately confirm whether it works there too.

Looking at https://developer.wordpress.org/reference/functions/set_transient/, serializing is not baked into Core when using a persistent object cache, so that confirms my concern. It only works reliably when the database is used, because the `*_option()` functions serialize.

And then again, even with serialization, there is the previous concern, since `WP_Error` does not implement https://www.php.net/manual/en/class.serializable.php.

Particularly with the object cache observation, I don't think it's safe to use serialization for this.

Reply:
The `Serializable` interface is for customizing how serialization and unserialization is done. You can see that a vanilla class with a `private` member is serialized and unserialized with its data intact: https://3v4l.org/DYOPX

Reply:
That's fair. What about the object cache concern though? I think that's still a problem.

Reply:
External object caches would need to be able to store any `mixed` value that you can store in WordPress's internal object cache, including _objects_, right? Examples from object caching plugins:

* [wp-redis](https://github.com/OllieJones/sqlite-object-cache/blob/215d9b228f94df12b0c3aff516ac5ebbff92ab2d/assets/drop-in/object-cache.php#L1464)
* [sqlite-object-cache](https://github.com/OllieJones/sqlite-object-cache/blob/215d9b228f94df12b0c3aff516ac5ebbff92ab2d/assets/drop-in/object-cache.php#L1464)
* [redis-cache](https://github.com/rhubarbgroup/redis-cache/blob/843617f6475f7916b0f8ca675704de88d5f7d20d/includes/object-cache.php#L2231-L2235)

Reply:
For [memcached-object-cache](https://github.com/Automattic/wp-memcached/blob/bb3b9f689dd99df66454b93ece34093556dc37f9/object-cache.php#L792) it passes the value directly to `Memcache`, but it turns out that `Memcache::set()` handles the serialization for you: https://www.php.net/manual/en/memcache.set.php

Reply:
[litespeed-cache](https://github.com/litespeedtech/lscache_wp/blob/d3e8158950e95a561aa4b64827714571b7e80614/src/object-cache.cls.php#L495-L497) passes the value directly to [`Redis::setEx()`](https://github.com/phpredis/phpredis/blob/faa4bc20868c76be4ecc4265015104a8adafccc4/redis.stub.php#L3110-L3121) which takes a `mixed` type. This function also is marked as one that [supports serialization](https://github.com/phpredis/phpredis/blob/faa4bc20868c76be4ecc4265015104a8adafccc4/serialize.list#L5).

Reply:
So I think that it's safe to say that if an object cache doesn't implement serialization for non-scalar values (including arrays not just objects) then it is fundamentally broken.

Reply:
I'm not so sure about this. I have never thought about storing any class instances in an option or object cache until this day. That's obviously not an argument not to use serialization, but probably many others haven't either.

Based on your research, I'm not necessarily disagreeing that external object caches should support this. But it's safe to say that it's not safe to rely on that they do :)

Since the effort to parse out the data from `WP_Error` into a raw array (and back) is minimal, why don't we simply go the safe route that we know will work?

Reply:
If there's a similar precedent in Core, I think it's a confirmation of what you're saying. But if not, we shouldn't rely on that any cache implementations should support this because it doesn't seem defined anywhere.

Reply:
Taking a step back though, we're probably arguing too much about this detail. I think there is a risk, but feel free to include it. If someone reports this later as a problem, we can still address it.

Reply:
Take a look at `fetch_feed()` such as used in the RSS widget. It stores an array that contains objects in it.

FWIW, I've been aware of storing class instances in transients and options since before I can remember. What's the difference with arrays? Both require serialization.

Reply:
I was googling for other examples, and I found a reference to Site Kit storing objects in transients: https://wpengine.com/resources/guide-to-transients-in-wordpress/

I can't find the current reference to that code in GitHub though.

Here's an example in the Transients documentation about storing a `WP_Query` instance in a transient: https://developer.wordpress.org/apis/transients/#complete-example

Reply:
Here's another example where an object is stored in a transient in `wp_version_check()`: https://github.com/WordPress/wordpress-develop/blob/2f654881e494424634d5821d1ef37c06edb8923a/src/wp-includes/update.php#L36-L73

Reply:
Objects are also stored in object cache in `WP_Term_Query`: https://github.com/WordPress/wordpress-develop/blob/2f654881e494424634d5821d1ef37c06edb8923a/src/wp-includes/class-wp-term-query.php#L879-L902

Reply:
Okay, you have me convinced. With this already done elsewhere in Core, this shouldn't be an issue, and the `WP_Error` class is basically always available, so there shouldn't be a problem unserializing it. 👍